### PR TITLE
to_upper and to_lower views

### PR DESCRIPTION
**Pull request by tloka**
Views on upper and lower case representations (requires char_concept).
Implemented using view::deep to ensure direct usage with ranges over char_concept.
Will later be used for case insensitive file extension validation.

### CODE REVIEW

File: test/snippet/range/view/to_case.cpp [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+{
+
+{
+//! [to_lower]
+std::string s{"CHanGED!"};
+std::vector<std::string> sv{"CHanGED!", "unchanged"};
+auto v1 = s | view::to_lower; 
+auto v2 = sv | view::to_lower; 
+
+debug_stream << "to_lower" << '\n' << "--------" << '\n';
```
Comment (smehringer) [inline test/snippet/range/view/to_case.cpp]: I would remove this as it does not help to understand view::to_lower ;)

File: test/snippet/range/view/to_case.cpp [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+debug_stream << v1 << '\n'; // => "changed!"
+debug_stream << v2 << '\n' << '\n'; // => ["changed!", "unchanged"]
```
Comment (smehringer) [inline test/snippet/range/view/to_case.cpp]:
```suggestion
debug_stream << v2 << '\n'; // => ["changed!", "unchanged"]
```

File: test/unit/range/view/view_to_lower_test.cpp [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    EXPECT_TRUE(std::ranges::BidirectionalRange<decltype(input_string)>);
+    EXPECT_TRUE(std::ranges::RandomAccessRange<decltype(input_string)>);
+    EXPECT_FALSE(std::ranges::View<decltype(input_string)>);
+    EXPECT_TRUE(std::ranges::SizedRange<decltype(input_string)>);
+    EXPECT_TRUE(std::ranges::CommonRange<decltype(input_string)>);
+    EXPECT_TRUE(const_iterable_concept<decltype(input_string)>);
+    EXPECT_TRUE((std::ranges::OutputRange<decltype(input_string), char>));
+
+    auto v1 = input_string | view::to_lower;
+    EXPECT_TRUE(std::ranges::InputRange<decltype(v1)>);
```
Comment (smehringer) [inline test/unit/range/view/view_to_lower_test.cpp]: I would rather do:
```cpp
EXPECT_EQ(std::ranges::InputRange<decltype(input_string)>, std::ranges::InputRange<decltype(input_string)>);
```
for properties that are preserved.
And
```cpp
EXPECT_EQ(std::ranges::InputRange<decltype(input_string)>, !std::ranges::InputRange<decltype(input_string)>);
```
(with `!`) for not preserved.

(Even if you might have copied this from somewhere else I think it is more clear).

File: test/unit/range/view/view_to_lower_test.cpp [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    EXPECT_TRUE(std::ranges::ForwardRange<decltype(v1)>);
+    EXPECT_TRUE(std::ranges::BidirectionalRange<decltype(v1)>);
+    EXPECT_TRUE(std::ranges::RandomAccessRange<decltype(v1)>);
+    EXPECT_TRUE(std::ranges::View<decltype(v1)>);
+    EXPECT_TRUE(std::ranges::SizedRange<decltype(v1)>);
+    EXPECT_TRUE(std::ranges::CommonRange<decltype(v1)>);
+    EXPECT_TRUE(const_iterable_concept<decltype(v1)>);
+    EXPECT_FALSE((std::ranges::OutputRange<decltype(v1), char>));
```
Comment (smehringer) [inline test/unit/range/view/view_to_lower_test.cpp]: You do not test that std::ranges::ContigousRange is not preserved, do you?

File: include/seqan3/range/view/to_lower.hpp [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+ *
+ * See the \link view view submodule documentation \endlink for detailed descriptions of the view properties.
+ *
+ * ### Example
+ * \snippet test/snippet/range/view/to_case.cpp to_lower
+ * \hideinitializer
+ */
+inline auto const to_lower = deep{std::view::transform([] (auto const in) noexcept
+{
+	static_assert(char_concept<remove_cvref_t<decltype(in)>>, "The value type of seqan3::view::to_lower must model the seqan3::char_concept.");
```
Comment (smehringer) [inline include/seqan3/range/view/to_lower.hpp]:
```suggestion
	static_assert(char_concept<remove_cvref_t<decltype(in)>>, 
	              "The value type of the innermost range must model the seqan3::char_concept.");
```
(line break if it is longer than 120 chars)

File: test/snippet/range/view/to_case.cpp [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+using namespace seqan3;
+
+int main()
+{
+
+{
+//! [to_lower]
+std::string s{"CHanGED!"};
+std::vector<std::string> sv{"CHanGED!", "unchanged"};
```
Comment (smehringer) [inline test/snippet/range/view/to_case.cpp]: In your example design the `!` is unchanged so it should be `{"CHanGED", "unchanged!"}` right? :nerd_face:

File: test/unit/range/view/view_to_lower_test.cpp [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+{
+
+    std::string input_string {"IAmADnaString"};
+    std::string cmp_string {"iamadnastring"};
+    std::string cmp2{"gnirtsandamai"};
+
+    std::vector<dna5> dna_vec {"AGGCGT"_dna5};
+    std::string cmp3{"aggcgt"};
+
+    std::vector<std::string> input_vec{"IAmADnaString", "IAmAProteinString"};
```
Comment (smehringer) [inline test/unit/range/view/view_to_lower_test.cpp]: Can you split this test into several smaller ones?

If something goes wrong it is easier to see in the googletest framework.
I would distinguish between basic, basic_deep and piping/combinability

File: test/unit/range/view/view_to_lower_test.cpp [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    std::vector<std::string> cmp4{"iamadnastring", "iamaproteinstring"};
+
+    // pipe notation string
+    std::string s(input_string | view::to_lower);
+    EXPECT_EQ(cmp_string, s);
+
+    // function notation string
```
Comment (smehringer) [inline test/unit/range/view/view_to_lower_test.cpp]:
```suggestion
    // custom conversion operator
```

File: include/seqan3/range/view/to_lower.hpp [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+ * See the \link view view submodule documentation \endlink for detailed descriptions of the view properties.
+ *
+ * ### Example
+ * \snippet test/snippet/range/view/to_case.cpp to_lower
+ * \hideinitializer
+ */
+inline auto const to_lower = deep{std::view::transform([] (auto const in) noexcept
+{
+	static_assert(char_concept<remove_cvref_t<decltype(in)>>, 
+				  "The value type of seqan3::view::to_lower must model the seqan3::char_concept.");
```
Comment (smehringer) [inline include/seqan3/range/view/to_lower.hpp]: fix indentation that I introduced :nerd_face:

File: test/unit/range/view/view_to_lower_test.cpp [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+#include <seqan3/range/concept.hpp>
+#include <seqan3/range/view/to_char.hpp>
+#include <seqan3/range/view/to_lower.hpp>
+#include <seqan3/std/ranges>
+
+using namespace seqan3;
+
+TEST(view_to_lower, basic)
+{
+
```
Comment (smehringer) [inline test/unit/range/view/view_to_lower_test.cpp]: remove new line :nail_care:

File: test/unit/range/view/view_to_upper_test.cpp [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+#include <seqan3/range/concept.hpp>
+#include <seqan3/range/view/to_char.hpp>
+#include <seqan3/range/view/to_upper.hpp>
+#include <seqan3/std/ranges>
+
+using namespace seqan3;
+
+TEST(view_to_upper, basic)
+{
+
```
Comment (smehringer) [inline test/unit/range/view/view_to_upper_test.cpp]: remove newline :nail_care:

File: include/seqan3/range/view/to_lower.hpp [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+ * |                                 |                                       |                                                    |
+ * | std::ranges::ViewableRange      | *required*                            | *guaranteed*                                       |
+ * | std::ranges::View               |                                       | *guaranteed*                                       |
+ * | std::ranges::SizedRange         |                                       | *preserved*                                        |
+ * | std::ranges::CommonRange        |                                       | *preserved*                                        |
+ * | std::ranges::OutputRange        |                                       | *lost*                                             |
+ * | seqan3::const_iterable_concept  |                                       | *preserved*                                        |
+ * |                                 |                                       |                                                    |
+ * | seqan3::reference_t             | seqan3::char_concept                  | *preserved*                                        |
```
Comment (h-2) [inline include/seqan3/range/view/to_lower.hpp]:
```suggestion
 * | seqan3::reference_t             | seqan3::char_concept                  | seqan3::remove_reference_t<seqan3::reference_t<urngt_>>  |
```

And adapt spacing of table separators

Comment (h-2) [reply include/seqan3/range/view/to_lower.hpp]: The reason here is that the return type is not the same, it is the underlying element type **minus `&`**

File: include/seqan3/range/view/to_upper.hpp [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+ * |                                 |                                       |                                                    |
+ * | std::ranges::ViewableRange      | *required*                            | *guaranteed*                                       |
+ * | std::ranges::View               |                                       | *guaranteed*                                       |
+ * | std::ranges::SizedRange         |                                       | *preserved*                                        |
+ * | std::ranges::CommonRange        |                                       | *preserved*                                        |
+ * | std::ranges::OutputRange        |                                       | *lost*                                             |
+ * | seqan3::const_iterable_concept  |                                       | *preserved*                                        |
+ * |                                 |                                       |                                                    |
+ * | seqan3::reference_t             | seqan3::char_concept                  | *preserved*                                        |
```
Comment (h-2) [inline include/seqan3/range/view/to_upper.hpp]: see above

File: include/seqan3/range/view/to_lower.hpp [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+ *
+ * See the \link view view submodule documentation \endlink for detailed descriptions of the view properties.
+ *
+ * ### Example
+ * \snippet test/snippet/range/view/to_case.cpp to_lower
+ * \hideinitializer
+ */
+inline auto const to_lower = deep{std::view::transform([] (auto const in) noexcept
```
[the remark this answers is not in the document]
Comment (tloka) [reply include/seqan3/range/view/to_lower.hpp]: Doesn't work, so I keep it as is.

File: test/snippet/range/view/to_case.cpp [hunk 15]
```diff
@@ -0,0 +1,38 @@
+#include <seqan3/io/stream/debug_stream.hpp>
+#include <seqan3/range/view/to_lower.hpp>
+#include <seqan3/range/view/to_upper.hpp>
+
+#include <vector>
+
+
```
Comment (eseiler) [inline test/snippet/range/view/to_case.cpp]: remove newline

**Pull request merged**